File: androidApp/src/main/java/cmunaro/github/analyticsdocumantationsample/android/home/HomeViewModel.kt
package cmunaro.github.analyticsdocumantationsample.android.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import cmunaro.github.analyticsdocumantationsample.analytics.domain.AnalyticsRepository
import cmunaro.github.analyticsdocumantationsample.android.utils.Async
import cmunaro.github.analyticsdocumantationsample.bank.domain.BankRepository
import cmunaro.github.analyticsdocumantationsample.events.BalanceDecreasedEvent
import cmunaro.github.analyticsdocumantationsample.events.BalanceIncreasedEvent
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn

class HomeViewModel(
    private val bankRepository: BankRepository,
    private val analyticsRepository: AnalyticsRepository
) : ViewModel() {
    val state = bankRepository.balance.map {
        HomeState(accountBalance = Async.Success(it))
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.Lazily,
        initialValue = HomeState()
    )

    fun onAdd1000Dollars() {
        runCatching {
            bankRepository.addMoney(1000.0)
        }.onSuccess {
            analyticsRepository.log(BalanceIncreasedEvent(delta = 1000.0))
        }
    }

    fun onSubtract1000Dollars() {
        runCatching {
            bankRepository.subtractMoney(1000.0)
        }.onSuccess {
            analyticsRepository.log(BalanceDecreasedEvent(delta = 1000.0))
        }
    }
}

File: iosApp/iosApp/HomeViewModel.swift
import SwiftUI
import Combine
import shared

@MainActor
class HomeViewModel: ObservableObject {
    private let bankRepository = DependenciesProviderHelper.shared.bankRepository
    private let analyticsRepository = DependenciesProviderHelper.shared.analyticsRepository
    @Published var homeState: HomeState

    init() {
        homeState = HomeState(accountBalance: 0.0)
        Task {
            try await bankRepository.balance
                .collect(collector: FlowCollector<Double>() { balance in
                    self.homeState = HomeState(accountBalance: balance)
                })
        }
    }

    func add1000Dollars() {
        do {
            try bankRepository.addMoney(amount: 1000.0)
            analyticsRepository.log(event: BalanceIncreasedEvent(delta:1000.0))
        } catch {}
    }

    func subtract1000Dollars() {
        do {
            try bankRepository.subtractMoney(amount: 1000.0)
            analyticsRepository.log(event: BalanceDecreasedEvent(delta:1000.0))
        } catch {}
    }
}
